YAML: create yaml test files for subroutines modal_aero_bcscavcoef_init,  modal_aero_bcscavcoef_get and and calc_1_impact_rate.

tangshuaiqi opened the pull request:
The initialize subroutine modal_aero_bcscavcoef_init and its child subroutine calc_1_impact_rate is only executed at timestep 0 without lchnk assigned. The test file generation code run will fail after the testing files are generated. I am keeping it for now as it already provides what we need but can be revised in the future.

Some other edits:
make a minor change on the template file beg.inp to assign function names.
adding yaml functions of writing 2d variables in int and float format
correct the error of lunerr in calc_1_impact_rate subroutine

Commit:
aero_model_wetdep: separate a subroutine to calculate surface fluxes from vertical integration

Commit:
aero_model_wetdep: clean the condition for lphase == 2 and revise some comments

Commit:
calc_1_impact_rate: move lunerr into the subroutine as local variable

Commit:
aero_model_wetdep: clean up code and comments in call wetdepa_v2

Commit:
aero_model_wetdep: separate subroutine calc_resusp_to_coarse

Commit:
aero_model_wetdep: separate subroutine apportion_sfc_flux_deep

Commit:
aero_model_wetdep: Clean up code related to sol_fact, remove code related to sscav_tuning

Commit:
aero_model_wetdep: clean and reorganize variable initiation

Commit:
aero_model_wetdep: separate subroutine index_ordering and simplify some index treatments

Commit:
aero_model_wetdep: separate subroutine define_act_frac

Commit:
aero_model_wetdep: clean up unused variables

Commit:
aero_model_wetdep: separate subroutine set_f_act_coarse

Commit:
aero_model_wetdep: remove the use of strt_loop and end_loop and simplify the code

Commit:
aero_model_wetdep: separate subroutine examine_prec_exist

Commit:
aero_model_wetdep: remove unused variables and chang single-digit vars into multi-digits

Commit:
aero_model_wetdep: remove unused input args

Commit:
aero_model_wetdep: reorganize arguments and add units

Commit:
aero_model_wetdep: change physpkg as input variables in aero_model_wetdep changed

Commit:
aero_model_wetdep: minor revision on some comments

Commit:
aero_model_wetdep: revise a comment to be more specific

Commit:
Add in, out, inout notation; remove more unneeded code

Commit:
Move kk loop through levels inside update_from_newcld

Commit:
Create new subroutine update_from_explmix to handle mixing substepping section of dropmixnuc.

Commit:
Remove unused code after last refactoring step.

Commit:
Modify ccncalc to utilize 1-col subroutines

Commit:
Remove non-1-col versions of subroutines since no longer used.

Commit:
Revert ccncalc to non-1col function to restore BFB.

Commit:
Combine code for qcldbrn

Commit:
Adding parameters and some restructure

Commit:
Fixed bug.

Commit:
YAML: create yaml test files for subroutines modal_aero_bcscavcoef_init,  modal_aero_bcscavcoef_get and and calc_1_impact_rate.
The initialize subroutine modal_aero_bcscavcoef_init and its child subroutine calc_1_impact_rate is only executed at timestep 0 without lchnk assigned. The location information in the test files are all zeros.
also make a minor change on the template file beg.inp

Commit:
aero_model_wetdep: add jaeronumb, jaeromass, jaerowater back

Commit:
aero_model_wetdep: explicitly list all variables used from modal_aero_data

Commit:
aero_model_wetdep: add comments to output argument of calc_sfc_flux

Commit:
aero_model_wetdep: change the treatment of switching mode order

Commit:
aero_model_wetdep: declare mam_prevap_resusp_optcc related number options to variables so it is easier to understand the meaning

tangshuaiqi commented:
All long-term tests passed:

comp_testall_yaml_bcscavcoef_mam4_org_v2_baselines: 9 tests
  PASS ERP_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS ERS.ne4pg2_oQU480.F2010.compy_intel
  PASS PEM_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS PET.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P1x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P32x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_Ly1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS.ne30pg2_EC30to60E2r2.F2010.compy_intel

Commit:
Merge pull request #90 from eagles-project/refactor-maint-2.0

merge the latest update of maint into branch

singhbalwinder reviewed (approved):
Looks good! It is a non-standard case for testing infrastructure so it is okay if the code fails at the end. It would be better to just comment-out or delete the unnecessary code (like lchnk, icolprnt etc.) as it will make the logic simpler. It is okay to leave it like this for this PR. We can revise it once we have a support for printing the init routines. BTW what error message do you get at the end?

tangshuaiqi commented:
@singhbalwinder The error message is below:

0: forrtl: severe (408): fort: (3): Subscript #1 of the array LCHUNKS has value 1 which is less than the lower bound of 97

Commit:
Remove extraneous comments.

Commit:
Create more parameter declarations for constants.

odiazib commented:
> @singhbalwinder The error message is below:
> 
> 0: forrtl: severe (408): fort: (3): Subscript #1 of the array LCHUNKS has value 1 which is less than the lower bound of 97

Is this error due to scavimptblnum and scavimptblvol have negative index? see [here ](https://github.com/eagles-project/e3sm_mam4_refactor/blob/26913eb0ffd0fb42cbaa0613f4ea0acec6524d93/components/eam/src/chemistry/modal_aero/aero_model.F90#L67).

odiazib commented on components/eam/src/chemistry/yaml/aero_model/f90_yaml/modal_aero_bcscavcoef_init_end_yml.f90:
```diff
@@ -0,0 +1,38 @@
+     !start writing data
+     !<add code for writing data here>
+     !call write_1d_output_var(unit_output, fld_name, dim, field, inp_out_str) !writes 1D variables of any dimension in the output python module
+     call write_2d_output_var(unit_output, 'scavimptblnum', size(scavimptblnum,1), ntot_amode, scavimptblnum) !writes 2D variables of any dimension in the output python module
```
`scavimptblnum` is defined [here](https://github.com/eagles-project/e3sm_mam4_refactor/blob/26913eb0ffd0fb42cbaa0613f4ea0acec6524d93/components/eam/src/chemistry/modal_aero/aero_model.F90#L69) as scavimptblnum(nimptblgrow_mind:nimptblgrow_maxd, ntot_amode) where nimptblgrow_mind is -7. But in write_2d_output_var indexing starts in 1. Is this an issue?

tangshuaiqi replied:
The two variables `scavimptblnum` and `scavimptblvol` have dimensions of (-7:12, ntot_amode) where -7:12 includes 0 (so the total length is 20, to answer your question in slack). Not sure how this type of indexing is treated in C++ but it could be an issue in testing file. One possible way is to change the dimensions of variables to 1:20 and revise the index accordingly (e.g., change `scavimptblvol(0,imode)` to `scavimptblvol(1-nimptblgrow_mind, imode)` ). @singhbalwinder do you have any suggestion on this issue?

singhbalwinder replied:
This is an interesting case, we need to add this capability since Fortran allows negative indices. Functions like `ubound` and `lbound` might be helpful here to identify the lower and upper bounds of an array within the `write_2d_output_var` subroutine.

odiazib replied:
in C++, we used 0 to 19 for the index.

tangshuaiqi replied:
Hi @singhbalwinder , I checked the `lbound` and `ubound` of variables that passed to `write_2d_output_var` and found that the bound indices actually changed. When I write out lbound and ubound from the `aero_model.F90` module, it is (-7, 12). But when I write out lbound and ubound from `write_2d_output_var`, it becomes (1, 20). So the original i/o functions actually write out the right order of the array.

odiazib commented on components/eam/src/chemistry/yaml/aero_model/modal_aero_bcscavcoef_init_input_ts_0.yaml:
```diff
@@ -0,0 +1,10 @@
+    dt:    0.00
+    dgnum_amode: [ 0.1100000000E-06, 0.2600000000E-07, 0.2000000000E-05, 0.5000000058E-07]
+    sigmag_amode: [ 0.1800000000E+01, 0.1600000000E+01, 0.1800000000E+01, 0.1600000024E+01]
+    specdens_amode: [ 0.1770000000E+04, 0.1797693135+309, 0.1797693135+309, 0.1000000000E+04]
```
what is this 0.1797693135+309 in `specdens_amode`?

tangshuaiqi replied:
looks like this is the mode/species that are not exist in the current model (it reads `specdens_amode(ll)` and `ll = lspectype_amode(1,imode)`, while `lspectype_amode` does not have number 2 and 3). I believe we can simply ignore it. @singhbalwinder do you have any comment on it?

odiazib commented on components/eam/src/chemistry/yaml/aero_model/f90_yaml/modal_aero_bcscavcoef_init_beg_yml.f90:
```diff
@@ -0,0 +1,81 @@
+        !< add code for writing data here>
+        call write_1d_var(unit_input, unit_output, 'dgnum_amode',ntot_amode, dgnum_amode)!write a single variable
+        call write_1d_var(unit_input, unit_output, 'sigmag_amode',ntot_amode,sigmag_amode)
+        call write_1d_var(unit_input, unit_output, 'specdens_amode',ntot_amode, specdens_amode)
```
`specdens_amode` has more than 4 values; [see here](https://github.com/eagles-project/e3sm_mam4_refactor/blob/26913eb0ffd0fb42cbaa0613f4ea0acec6524d93/components/eam/src/chemistry/modal_aero/modal_aero_data.F90#L197)

tangshuaiqi replied:
You are right. I'll correct it

odiazib commented on components/eam/src/chemistry/yaml/aero_model/modal_aero_bcscavcoef_get_input_ts_355.yaml:
```diff
@@ -0,0 +1,11 @@
+    dt:    0.00
+    imode: [         1]
+    ncol: [         4]
+    isprx: [         T]
```
Could you please change this T to true in `isprx `?

odiazib replied:
T or true does not work in Skywalker, we need an integer, i.e.,  `T -> 1` or `F-> 0`

tangshuaiqi replied:
currently I am using a naive way to change it (see below) as `int(isprx)` doesn't work (not sure why). we may want to change the write_var subroutine if Skywalker doesn't read logical vars. @singhbalwinder any comments?


`    

       if (isprx(yaml%col_print, yaml%lev_print)) then
                logic_int = 1
        else
                logic_int = 0
        endif
        call write_var(unit_input, unit_output, 'isprx', logic_int)`

singhbalwinder replied:
If you can, would you please try changing the `write` statement in `components/eam/src/chemistry/utils/yaml_input_file_io.F90` file to print the logical like an integer (replace `L` with `I`):

```
write(unit_input,'(3A,I10,A)')'    ',trim(adjustl(fld_name)),': [', field,']'
```

This will change F->0 but the T can be a -1 or 1. We can change the above statement to print positive 1 using `abs`:
```
write(unit_input,'(3A,I10,A)')'    ',trim(adjustl(fld_name)),': [', abs(field),']'
```

You can make this change in all the subroutines ending with `_logical`.

singhbalwinder replied:
`abs` might not work as it is a logical, not integer. There is another Fortran intrinsic called `merge` that can be used here like:
```
write(unit_input,'(3A,I10,A)')'    ',trim(adjustl(fld_name)),': [', merge(1,0,field),']'
```

[Here ](https://gcc.gnu.org/onlinedocs/gcc-4.4.7/gfortran/MERGE.html)is some explanation of the merge function. I haven't tried it but I hope it works otherwise, we can use an explicit `if` condition.

tangshuaiqi replied:
revised as suggested

odiazib commented on components/eam/src/chemistry/yaml/aero_model/modal_aero_bcscavcoef_get_input_ts_355.yaml:
```diff
@@ -0,0 +1,11 @@
+    ncol: [         4]
+    isprx: [         T]
+    dgn_awet: [ 0.1522452427E-06, 0.4627511718E-07, 0.1886138608E-05, 0.9911179496E-07]
+    dgnum_amode: [ 0.1100000000E-06, 0.2600000000E-07, 0.2000000000E-05, 0.5000000058E-07]
```
Could you add `scavimptblnum` and `scavimptblvol`?

Commit:
Create even more parameters from constants, organize variable declarations and remove unneeded ones.

Commit:
Move icol loop within loadaer, modify other subroutines accordingly
Change order of indices in qcldbrn
Additional parameter specification and reformatting

Commit:
aero_model_wetdep: add ncol in calc_sfc_flux calls to avoid array mismatch error

Commit:
Merge pull request #73 from eagles-project/tangshuaiqi/aero_model_wetdep

refactor aero_model_wetdep

Commit:
YAML: correct the dimension of specdens_amode; change logical variable writeout to integer; add scavimptblnum and scavimptblvol in modal_aero_bcscavcoef_get input file

Commit:
YAML: add missing output variables

odiazib commented on components/eam/src/chemistry/yaml/aero_model/f90_yaml/modal_aero_bcscavcoef_get_beg_yml.f90:
```diff
@@ -0,0 +1,80 @@
+
+  !YAML file input generation code- DO NOT PORT to C++
+  !print all inputs one-by-one at column "yaml%col_print"
+  yaml%col_print = icolprnt(y_lchnk) !column to write data(**remove these if generating data for a dependent subroutines**)
```
what is `yaml%col_print` ? , is this 4, i.e., `ncol`?

odiazib commented on components/eam/src/chemistry/yaml/aero_model/modal_aero_bcscavcoef_get_output_ts_355.py:
```diff
@@ -0,0 +1,22 @@
+input.imode=[[         1],]
+input.ncol=[[         4],]
+input.isprx=[[         T],]
+input.dgn_awet=[[ 0.1522452427E-06, 0.4627511718E-07, 0.1886138608E-05, 0.9911179496E-07,],]
```
if there are only 4 cols. Can we save `dgn_awet(pcols,pver,ntot_amode)  ` for 4 cols, i.e., `dgn_awet(:,53,:) `.

odiazib commented on components/eam/src/chemistry/yaml/aero_model/modal_aero_bcscavcoef_get_output_ts_355.py:
```diff
@@ -0,0 +1,22 @@
+input.dgnum_amode=[[ 0.1100000000E-06, 0.2600000000E-07, 0.2000000000E-05, 0.5000000058E-07,],]
+# Output data is stored here.
+output = Object()
+output.scavcoefnum=[[ 0.6254861968E-03],]
```
Can we save `scavcoefnum` and `scavcoefvol` for the ncol?. In this case,  this output will be a vector of 4 values.

tangshuaiqi replied:
yes we can save the variables for all ncols. Just wondering, do you want ncols only for this subroutine or for all tests in the future?

odiazib replied:
only for this subroutine.Thanks!

tangshuaiqi replied:
test files updated.

Commit:
YAML: change writeout of logical variables to integer format. add modal_aero_bcscavcoef_get output variables in icol dimension as requested by C++ team

singhbalwinder committed:
code indentation

singhbalwinder committed:
Merge pull request #87 from eagles-project/brigaudet/refactor-dropmixnuc_step3

Step 3 of dropmixnuc refactor

odiazib reviewed (approved):
Thanks! @tangshuaiqi

Commit:
YAML: change the yaml i/o subroutines with lbound and ubound. input variables dim1/dim2 are now not used but still kept to prevent failure of previous code

tangshuaiqi commented:
I have changed the yaml i/o subroutines (in `yaml_input_file_io.F90`) with lbound and ubound (see commit [2244433]). The results are still the same as an array with pre-defined indices (e.g., from -7 to 12) will change to natural indices (e.g., 1 to 20) in the subroutine. After the changes, the input variables `dim1` and `dim2` in `write_2d_output_var_real(unit_output, fld_name, dim1, dim2, field, inp_out_str) `are not used but are still kept to avoid failure of previous code.

Commit:
solve a conflict in merge

Commit:
a minor fix

tangshuaiqi commented:
All tests passed. merge into maint
  PASS ERP_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS ERS.ne4pg2_oQU480.F2010.compy_intel
  PASS PEM_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS PET.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P1x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P32x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_Ly1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS.ne30pg2_EC30to60E2r2.F2010.compy_intel
